flask_menu/ext.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Flask-Menu
+# Copyright (C) 2013, 2014, 2015, 2017 CERN
+# Copyright (C) 2017 Marlin Forbes
+# Copyright (C) 2023 Graz University of Technology.
+#
+# Flask-Menu is free software; you can redistribute it and/or modify
+# it under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Ext."""
+
+from flask import current_app, g
+
+from .menu import MenuRoot
+
+
+class Menu:
+    """Flask extension implementation."""
+
+    def __init__(self, app=None):
+        """Initialize Menu extension."""
+        if app is not None:
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Initialize a Flask application."""
+        self.app = app
+        # Follow the Flask guidelines on usage of app.extensions
+        if not hasattr(app, "extensions"):
+            app.extensions = {}
+        if "menu" in app.extensions:
+            raise RuntimeError("Flask application is already initialized.")
+        app.extensions["menu"] = MenuRoot("", None)
+
+        app.context_processor(lambda: {"current_menu": Menu.root()})
+
+        @app.url_value_preprocessor
+        def url_preprocessor(route, args):
+            """Store the current route endpoint and arguments."""
+            g._menu_kwargs = args
+            g._menu_route = route
+
+    @staticmethod
+    def root():
+        """Return a root entry of current application's menu."""
+        return current_app.extensions["menu"]
```

Comment:
```suggestion
class Menu:
    """Flask extension implementation."""

    def __init__(self, app=None, root=None):
        """Initialize Menu extension."""
        slef.root = root or MenuRoot("", None)
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        """Initialize a Flask application."""
        self.app = app
        # Follow the Flask guidelines on usage of app.extensions
        if not hasattr(app, "extensions"):
            app.extensions = {}
        if "menu" in app.extensions:
            raise RuntimeError("Flask application is already initialized.")

        app.context_processor(lambda: {"current_menu": Menu.root()})

        @app.url_value_preprocessor
        def url_preprocessor(route, args):
            """Store the current route endpoint and arguments."""
            g._menu_kwargs = args
            g._menu_route = route

    @staticmethod
    def root():
        """Return a root entry of current application's menu."""
        return current_app.extensions["menu"].root
```

major: (after reviewing the usage in `invenio-theme`) this way we could decouple initialization and usage

Reply:
did you forgot the `app.extensions["menu"] = self` line in `init_app`?

Reply:
yes, I've been trying to edit it for a couple of minutes 🤦

Reply:
I think I introduced a naming clash, since there's now a `self.root` (passed to `__init__`) and the `Menu.root` static method. Could we drop the static method and adjust the proxy maybe?

Reply:
i am not sure i have to test that. i am little bit lost in the app context jungle